contracts/cw1-subkeys/src/contract.rs

```diff
@@ -102,18 +127,47 @@ pub fn handle_decrease_allowance<S: Storage, A: Api, Q: Querier, T>(
 where
     T: Clone + fmt::Debug + PartialEq + JsonSchema,
 {
-    // placeholder to remove warnings
-    let _ = (deps, env, spender, amount, expires);
-
-    // TODO
-    // look at cw20_base::contract::DecreaseAllowance
-    // if something present, we subtract from the balance. underflow is okay, removes that denom
-    // (eg. if allowance was [5 ETH, 2 BTC] and I decrease by 4 BTC, new allowance is [5 ETH].
-    // if final balance has no denoms, remove the allowance entry.
-    //
-    // as with handle_increase_allowance,
-    // expires=None => leave expiration at previous state, expires=Some(x) => set expires to new value x
-    panic!("unimplemented")
+    let cfg = config_read(&deps.storage).load()?;
+    let spender_raw = &deps.api.canonical_address(&spender)?;
+    let owner_raw = &deps.api.canonical_address(&env.message.sender)?;
+
+    if !cfg.is_admin(&owner_raw) {
+        return Err(StdError::unauthorized());
+    }
+    if spender_raw == owner_raw {
+        return Err(StdError::generic_err("Cannot set allowance to own account"));
+    }
+
+    let allowance = allowances(&mut deps.storage).update(spender_raw.as_slice(), |allow| {
+        if allow.is_none() {
+            // Fail fast
+            return Err(StdError::generic_err("No allowance for this account"));
+        }
+        let mut allowance = allow.unwrap();
+
+        if let Some(exp) = expires {
+            allowance.expires = exp;
+        }
+
+        allowance.balance = allowance.balance.sub(amount.clone())?; // Fails if no tokens
+        Ok(allowance)
+    })?;
+    if allowance.balance.is_empty() {
+        allowances(&mut deps.storage).remove(spender_raw.as_slice());
```

Comment:
Nice way to handle this without a second read.

You could also avoid update and just load, modify, and decide to save or remove.

Reply:
Yes. I'll change it if you want, but I like update: it prevents you from forgetting to save. I also like its functional style, sort of.

Reply:
It's okay either way. I do like update myself and I think you use it quite well here, especially using the return value. I think you could optimize, but feel this is more legible, so feel free to keep it.